climada/hazard/centroids/centr.py

```diff
@@ -79,280 +52,215 @@ class Centroids():
         """
 
-        self.lat = lat if lat is not None else np.array([])
-        self.lon = lon if lon is not None else np.array([])
-        self.geometry = geometry if geometry is not None else gpd.GeoSeries(crs=DEF_CRS)
-        self.meta = meta if meta is not None else dict()
-        self.area_pixel = area_pixel if area_pixel is not None else np.array([])
-        self.on_land = on_land if on_land is not None else np.array([])
-        self.region_id = region_id if region_id is not None else np.array([])
-        self.elevation = elevation if elevation is not None else np.array([])
-        self.dist_coast = dist_coast if dist_coast is not None else np.array([])
-
-    def check(self):
-        """Check integrity of stored information.
-
-        Checks that either `meta` attribute is set, or `lat`, `lon` and `geometry.crs`.
-        Checks sizes of (optional) data attributes."""
-        n_centr = self.size
-        for var_name, var_val in self.__dict__.items():
-            if var_name in self.vars_check:
-                if var_val.size > 0 and var_val.size != n_centr:
-                    raise ValueError(f'Wrong {var_name} size: {n_centr} != {var_val.size}.')
-        if self.meta:
-            for name in ['width', 'height', 'crs', 'transform']:
-                if name not in self.meta.keys():
-                    raise ValueError('Missing meta information: %s' % name)
-            xres, xshear, _xoff, yshear, yres, _yoff = self.meta['transform'][:6]
-            if xshear != 0 or yshear != 0:
-                raise ValueError('Affine transformations with shearing components are not '
-                                 'supported.')
-            if yres > 0 or xres < 0:
-                raise ValueError('Affine transformations with positive y-orientation '
-                                 'or negative x-orientation are not supported.')
-
-    def equal(self, centr):
+        attr_dict = {
+            'geometry': gpd.points_from_xy(lon, lat, crs=crs),
+            'region_id': region_id,
+            'on_land': on_land
+        }
+        self.gdf = gpd.GeoDataFrame(data=attr_dict)
+
+        if isinstance(region_id, str):
+            LOGGER.info(f'Setting region id to {region_id} level.')
+            self._set_region_id(level=region_id, overwrite=True)
+        if isinstance(on_land, str):
+            LOGGER.info(f'Setting on land from {on_land} source.')
+            self._set_on_land(source=on_land, overwrite=True)
+
+    @property
+    def lat(self):
+        """ Return latitudes """
+        return self.gdf.geometry.y.values
+
+    @property
+    def lon(self):
+        """ Return longitudes """
+        return self.gdf.geometry.x.values
+
+    @property
+    def geometry(self):
+        """ Return the geometry """
+        return self.gdf['geometry']
+
+    @property
+    def on_land(self):
+        """ Get the on_land property """
+        if self.gdf.on_land.isna().all():
+            return None
```

Comment:
I think checking all values for NaN can be very costly. We should instead check if the `on_land` column exists at all, and return `None` if it does not.

Reply:
I disagree. I think it makes more sense to have a consistent object with the same variables always defined in the geodataframe. The operation is not particularly costly. But if you have a better suggestion for handling this shoot.

Reply:
If the gdf is (intended to be) never modified once the class is built, [as you state](https://github.com/CLIMADA-project/climada_python/pull/787#discussion_r1440297148), then we can simply set our own logic. I think it's very wasteful to *always* check *all* values of `on_land` if this method is called, and it seems like a clear antipattern to me. Simply not setting the `on_land` column in the init should be safe, so I stand by my initial proposal.

Reply:
I think it is still useful to have a coherent dataframe, in particular it makes saving and loading much simpler.